website/addons/base/views.py

```diff
@@ -201,19 +207,79 @@ def create_waterbutler_log(payload, **kwargs):
     return {'status': 'success'}
 
 
+def get_or_start_render(file_guid, extra, start_render=True):
+    file_guid.maybe_version(**extra)
+
+    try:
+        file_guid.enrich()
+
+        return codecs.open(file_guid.cache_path, 'r', 'utf-8').read()
+    except exceptions.AddonEnrichmentError as error:
+        return error.renderable_error
+    except IOError:
+        if start_render:
+            # Start rendering job if requested
+            build_rendered_html(file_guid.download_url, file_guid.cache_path, file_guid.temp_path)
+    return None
+
+
 @must_be_valid_project
-def get_waterbutler_render_url(**kwargs):
-    provider = request.args.get('provider')
+@must_be_contributor_or_public
+def addon_view_or_download_file(auth, path, provider, **kwargs):
+    extras = request.args.to_dict()
+    mode = extras.pop('action', 'view')
     node = kwargs.get('node') or kwargs['project']
 
     node_addon = node.get_addon(provider)
 
-    if not node_addon:
+    if not path or not node_addon:
         raise HTTPError(httplib.BAD_REQUEST)
 
-    try:
-        url = node_addon.get_waterbutler_render_url(**request.args.to_dict())
-    except TypeError:
+    if not path.startswith('/'):
+        path = '/' + path
+
+    file_guid, created = node_addon.find_or_create_file_guid(path)
```

Comment:
Would be helpful to add a base WaterButler addon class that defines `find_or_create_file_guid` as abstract.